### PR TITLE
Generate point clouds on device

### PR DESCRIPTION
**Pull request by aprokop**
Generating large point clouds (like 100M) on host takes about half a minute (on Frontier), which makes it challenging to run multiple benchmarks. This PR does generation on the device. In addition, this PR fixes point generation for sphere and hollow sphere to produce uniform distributions and work better for high dimensions.

The generation is done in batched mode. I compared generating a certain number of random floats on A100 using different batches using the following driver:
[kokkos_vs_std_random.cpp](https://github.com/user-attachments/files/15537921/kokkos_vs_std_random.txt).

There's not much difference once one goes beyond batch 4:
```
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
BM_std_random/100                     7.85 us         7.85 us        89068
BM_std_random/10000                   56.6 us         56.6 us        12132
BM_std_random/1000000                 4582 us         4582 us          152
BM_std_random/100000000             457032 us       456937 us            2

BM_kokkos_random/100/1                13.4 us         13.4 us        40553
BM_kokkos_random/10000/1              13.3 us         13.3 us        52073
BM_kokkos_random/1000000/1             146 us          146 us         4831
BM_kokkos_random/100000000/1         13616 us        13616 us           52

BM_kokkos_random/100/4                12.1 us         12.1 us        57673
BM_kokkos_random/10000/4              13.7 us         13.7 us        51221
BM_kokkos_random/1000000/4            68.3 us         68.3 us        10250
BM_kokkos_random/100000000/4          4465 us         4462 us          158

BM_kokkos_random/100/16               13.3 us         13.3 us        52455
BM_kokkos_random/10000/16             16.1 us         16.1 us        43482
BM_kokkos_random/1000000/16           40.4 us         40.4 us        17361
BM_kokkos_random/100000000/16         3964 us         3963 us          177

BM_kokkos_random/100/64               18.3 us         18.3 us        38187
BM_kokkos_random/10000/64             26.6 us         26.6 us        26283
BM_kokkos_random/1000000/64           51.8 us         51.8 us        13491
BM_kokkos_random/100000000/64         5250 us         5249 us          133

BM_kokkos_random/100/256              1.52 us         1.52 us       462197
BM_kokkos_random/10000/256            44.1 us         44.1 us        15880
BM_kokkos_random/1000000/256          69.5 us         69.5 us        10059
BM_kokkos_random/100000000/256        6434 us         6432 us          110

BM_kokkos_random/100/1024             1.51 us         1.51 us       463793
BM_kokkos_random/10000/1024            119 us          119 us         5907
BM_kokkos_random/1000000/1024          230 us          230 us         3052
BM_kokkos_random/100000000/1024       5408 us         5408 us          129

BM_kokkos_random/100/4096             1.51 us         1.51 us       463578
BM_kokkos_random/10000/4096            439 us          439 us         1596
BM_kokkos_random/1000000/4096          872 us          872 us          802
BM_kokkos_random/100000000/4096       4397 us         4396 us          157
```
I chose 8, on the lower end, given that we are going to generate multi-dimensional points, which means `n * DIM`.

I validated the produced clouds by dumping them into VTK and visualizing.

### CODE REVIEW

**Comment by aprokop** [comment]
Here are the performance results:

### master (da4fb8bf) 
[master_da4fb8bf.log](https://github.com/user-attachments/files/15537968/master_da4fb8bf.log)

```
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
BM_filled_box/100              0.015 ms        0.015 ms        45426
BM_filled_box/1000             0.039 ms        0.039 ms        18009
BM_filled_box/10000            0.270 ms        0.270 ms         2588
BM_filled_box/100000            2.56 ms         2.56 ms          274
BM_filled_box/1000000           24.8 ms         24.8 ms           28
BM_filled_box/10000000           303 ms          303 ms            2
BM_hollow_box/100              0.015 ms        0.015 ms        47096
BM_hollow_box/1000             0.032 ms        0.032 ms        21959
BM_hollow_box/10000            0.202 ms        0.202 ms         3431
BM_hollow_box/100000            1.86 ms         1.86 ms          379
BM_hollow_box/1000000           17.7 ms         17.7 ms           40
BM_hollow_box/10000000           222 ms          222 ms            3
BM_filled_sphere/100           0.018 ms        0.018 ms        39861
BM_filled_sphere/1000          0.067 ms        0.067 ms        10422
BM_filled_sphere/10000         0.682 ms        0.682 ms         1027
BM_filled_sphere/100000         6.73 ms         6.73 ms          104
BM_filled_sphere/1000000        66.3 ms         66.3 ms           10
BM_filled_sphere/10000000        732 ms          732 ms            1
BM_hollow_sphere/100           0.017 ms        0.017 ms        41022
BM_hollow_sphere/1000          0.056 ms        0.056 ms        12549
BM_hollow_sphere/10000         0.464 ms        0.464 ms         1513
BM_hollow_sphere/100000         4.49 ms         4.49 ms          156
BM_hollow_sphere/1000000        44.3 ms         44.3 ms           16
BM_hollow_sphere/10000000        485 ms          485 ms            2
```

### This branch (serial) 
[points_device_serial.log](https://github.com/user-attachments/files/15537972/points_device_serial.log)

```
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
BM_filled_box/100              0.008 ms        0.008 ms        82958
BM_filled_box/1000             0.024 ms        0.024 ms        28931
BM_filled_box/10000            0.244 ms        0.244 ms         2877
BM_filled_box/100000            2.38 ms         2.38 ms          294
BM_filled_box/1000000           23.8 ms         23.8 ms           29
BM_filled_box/10000000           258 ms          258 ms            2
BM_hollow_box/100              0.008 ms        0.008 ms        84657
BM_hollow_box/1000             0.022 ms        0.022 ms        32099
BM_hollow_box/10000            0.203 ms        0.203 ms         3459
BM_hollow_box/100000            1.98 ms         1.98 ms          353
BM_hollow_box/1000000           19.8 ms         19.8 ms           36
BM_hollow_box/10000000           211 ms          211 ms            3
BM_filled_sphere/100           0.009 ms        0.009 ms        74673
BM_filled_sphere/1000          0.056 ms        0.056 ms        12418
BM_filled_sphere/10000         0.568 ms        0.568 ms         1233
BM_filled_sphere/100000         5.54 ms         5.54 ms          126
BM_filled_sphere/1000000        55.6 ms         55.6 ms           13
BM_filled_sphere/10000000        600 ms          600 ms            1
BM_hollow_sphere/100           0.009 ms        0.009 ms        77972
BM_hollow_sphere/1000          0.028 ms        0.028 ms        24659
BM_hollow_sphere/10000         0.268 ms        0.268 ms         2609
BM_hollow_sphere/100000         2.63 ms         2.62 ms          267
BM_hollow_sphere/1000000        26.2 ms         26.2 ms           27
BM_hollow_sphere/10000000        283 ms          283 ms            2
```

### This branch (device, A100) 
[points_device_a100.log](https://github.com/user-attachments/files/15537974/points_device_a100.log)

```
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
BM_filled_box/100               28.3 ms         28.3 ms           23
BM_filled_box/1000              28.1 ms         28.1 ms           25
BM_filled_box/10000             28.1 ms         28.1 ms           25
BM_filled_box/100000            28.1 ms         28.1 ms           25
BM_filled_box/1000000           29.8 ms         29.8 ms           24
BM_filled_box/10000000          31.0 ms         31.0 ms           23
BM_hollow_box/100               28.1 ms         28.1 ms           25
BM_hollow_box/1000              28.1 ms         28.1 ms           25
BM_hollow_box/10000             28.0 ms         28.0 ms           25
BM_hollow_box/100000            28.1 ms         28.1 ms           25
BM_hollow_box/1000000           29.4 ms         29.4 ms           23
BM_hollow_box/10000000          31.1 ms         31.1 ms           22
BM_filled_sphere/100            28.0 ms         28.0 ms           25
BM_filled_sphere/1000           28.4 ms         28.4 ms           25
BM_filled_sphere/10000          28.1 ms         28.1 ms           25
BM_filled_sphere/100000         28.1 ms         28.1 ms           25
BM_filled_sphere/1000000        50.4 ms         30.3 ms           24
BM_filled_sphere/10000000       34.6 ms         34.6 ms           20
BM_hollow_sphere/100            28.3 ms         28.2 ms           25
BM_hollow_sphere/1000           28.0 ms         28.0 ms           25
BM_hollow_sphere/10000          28.0 ms         28.0 ms           25
BM_hollow_sphere/100000         28.0 ms         28.0 ms           25
BM_hollow_sphere/1000000        29.3 ms         29.3 ms           24
BM_hollow_sphere/10000000       30.9 ms         30.9 ms           23
```

**Comment by aprokop** [comment]
As one can see, generating on the device does not really depend on the size, as most of the time is taken by grabbing the state from a random pool. I looked into it, and I still don't have any idea why it is so expensive, when it really should take no time at all. Still, 0.03 seconds is good enough for me.

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp [hunk 1]
```diff
@@ -49,160 +47,163 @@ inline PointCloudType to_point_cloud_enum(std::string const &str)
...
+    // For 3D, the order is
+    // Indices: 0  1  2  3  4  5  6  7  8  9
+    // Axes   : 0- 0+ 1- 1+ 2- 2+ 0- 0+ 1- 1+
+    int axis = (i / 2) % DIM;
+    for (int d = 0; d < DIM; ++d)
+    {
+      if (d != axis)
+        p[d] = random();
+      else if (i % 2 == 0)
+        p[d] = -1;
```
Comment (masterleinad) [inline benchmarks/utils/ArborXBenchmark_PointClouds.hpp:108]: This was previously `half_edge`.

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp [hunk 2]
```diff
@@ -49,160 +47,163 @@ inline PointCloudType to_point_cloud_enum(std::string const &str)
...
       for (int d = 0; d < DIM; ++d)
       {
-        double v = random();
-        p[d] = v;
-        norm += v * v;
+        p[d] = random();
+        norm += p[d] * p[d];
       }
-      norm = std::sqrt(norm);
+      norm = Kokkos::sqrt(norm);
```
Comment (masterleinad) [inline benchmarks/utils/ArborXBenchmark_PointClouds.hpp:130]: You could only take the square root at the end once and compare to the squared radius instead.

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp [hunk 3]
```diff
@@ -49,160 +47,163 @@ inline PointCloudType to_point_cloud_enum(std::string const &str)
...
-      norm += v[d] * v[d];
+      p[d] = random();
+      norm += p[d] * p[d];
     }
-    norm = std::sqrt(norm);
-
+    norm = Kokkos::sqrt(norm);
     for (int d = 0; d < DIM; ++d)
-      random_points(i)[d] = radius * v[d] / norm;
+      p[d] /= norm;
```
Comment (masterleinad) [inline benchmarks/utils/ArborXBenchmark_PointClouds.hpp:150]: Maybe retry if we hit the origin?

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp (hunk 1 above)
Comment (aprokop) [reply benchmarks/utils/ArborXBenchmark_PointClouds.hpp:108]: I now scale only in the loop, outside of the specific point generation functor.

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp (hunk 3 above)
Comment (aprokop) [reply benchmarks/utils/ArborXBenchmark_PointClouds.hpp:150]: I'm changing this code as it does not generate uniformly distributed points on the sphere.

File: benchmarks/utils/ArborXBenchmark_PointClouds.hpp (hunk 2 above)
Comment (aprokop) [reply benchmarks/utils/ArborXBenchmark_PointClouds.hpp]: Done.

**Review by aprokop — COMMENTED** [review]
The sphere algorithms are bad for high dimensions, where the volume of the sphere is insignificant portion of the encompassing box (https://en.wikipedia.org/wiki/Volume_of_an_n-ball). Need to switch them to something else.

**Comment by aprokop** [comment]
Following [this](https://blogs.sas.com/content/iml/2016/04/06/generate-points-uniformly-in-ball.html) blog

> If $Y$ is drawn from the uncorrelated multivariate normal distribution, then $S = Y / \|Y\|$ has the uniform distribution on the unit $d$-sphere. Multiplying $S$ by $U^{1/d}$, where $U$ has the uniform distribution on the unit interval $(0,1)$, creates the uniform distribution in the unit $d$-dimensional ball.

**Pull request merged**